WalletWasabi.Gui/Program.cs

```diff
@@ -116,25 +113,25 @@ private static async void AppMainAsync(string[] args)
 			}
 			catch (Exception ex)
 			{
-				if (!(ex is OperationCanceledException))
-				{
-					Logger.LogCritical(ex);
-					Global.CrashReporter.SetException(ex);
-				}
-
-				await DisposeAsync();
+				var criticalException = ex is OperationCanceledException ? null : ex;
 
 				// There is no other way to stop the creation of the WasabiWindow.
-				Environment.Exit(1);
+				await TerminateApplicationAsync(criticalException);
 			}
 		}
 
-		private static async Task DisposeAsync()
+		private static async Task TerminateApplicationAsync(Exception? criticalException = null)
 		{
-			var disposeGui = MainWindowViewModel.Instance is { };
-			if (disposeGui)
+			if (criticalException is { })
+			{
+				Logger.LogCritical(criticalException);
+				Global?.CrashReporter?.SetException(criticalException);
+			}
+
+			var mainViewModel = MainWindowViewModel.Instance;
+			if (mainViewModel is { })
 			{
-				MainWindowViewModel.Instance.Dispose();
+				mainViewModel.Dispose();
```

Comment:
Why not to dispose that in `AppMainAsync`? 

I mean:

```csharp
MainWindowViewModel.Instance = new MainWindowViewModel(Global.Network, Global.UiConfig, Global.WalletManager, statusBarViewModel, IoC.Get<IShell>());
```

to

```csharp
using var mainWindowViewModel = new MainWindowViewModel(Global.Network, Global.UiConfig, Global.WalletManager, statusBarViewModel, IoC.Get<IShell>());
MainWindowViewModel.Instance = mainWindowViewModel;

await Global.InitializeNoWalletAsync();

mainWindowViewModel.Initialize(Global.Nodes.ConnectedNodes);
```

Reply:
AppMainAsync returns after the initialization and passes the execution to the "Avalonia UI message pump".